src/ILVerification/src/InterfaceVerifier.cs

```diff
@@ -0,0 +1,157 @@
+                    else
+                    {
+                        InterfaceVerificationResult?.Invoke(new InterfaceVerificationResult()
+                        {
+                            Type = _typeDefinitionHandle,
+                            Error = new InterfaceVerificationErrorArgs()
+                            {
+                                Code = VerifierError.InterfaceImplHasDuplicate,
+                                TokenClass = _module.MetadataReader.GetToken(_typeDefinitionHandle),
+                                TokenInterface = _module.MetadataReader.GetToken(interfaceHandle)
+                            },
+                            Message = string.Format(_stringResourceManager.Value.GetString(VerifierError.InterfaceImplHasDuplicate.ToString(), CultureInfo.InvariantCulture), interfaceType.ToString())
+                        });
+                    }
+                }
+
+                // Looks for missing method implementation
+                foreach (InterfaceMetadataObjects interfaceMetadataObjects in implementedInterfaces.Distinct())
+                {
+                    foreach (MethodDesc method in interfaceMetadataObjects.DefType.GetAllMethods())
+                    {
+                        if (!IsImplemented(type, method))
+                        {
+                            InterfaceVerificationResult?.Invoke(new InterfaceVerificationResult()
+                            {
+                                Type = _typeDefinitionHandle,
+                                Error = new InterfaceVerificationErrorArgs()
+                                {
+                                    Code = VerifierError.InterfaceMethodNotImplemented,
+                                    TokenClass = _module.MetadataReader.GetToken(_typeDefinitionHandle),
+                                    TokenInterface = _module.MetadataReader.GetToken(interfaceMetadataObjects.InterfaceImplementationHandle)
+                                },
+                                Message = string.Format(_stringResourceManager.Value.GetString(VerifierError.InterfaceMethodNotImplemented.ToString(), CultureInfo.InvariantCulture), interfaceMetadataObjects.DefType.ToString(), method.ToString())
+                            });
+                        }
+                    }
+                }
+            }
+        }
+
+        // Search method matches name and sign recursively        
+        private bool IsImplemented(EcmaType type, MethodDesc interfaceMethod)
+        {
+            MethodDesc result = null;
+            TypeDesc currentType = type;
+            do
+            {
+                result = FindMethod(type, interfaceMethod);
+                currentType = currentType.BaseType;
+            }
+            while (result == null && currentType != null);
+
+            return result != null;
+        }
+
+        private MethodDesc FindMethod(EcmaType currentType, MethodDesc interfaceMethod)
```

Comment:
Looking at PE verifier source code I see two check https://github.com/lewischeng-ms/sscli/blob/master/clr/src/md/compiler/mdvalidator.cpp#L2581 : 
1) Search for virtual method name/sign. I added IL test for this case and it's ok
2) Check among MethodImpls. I want to write code also for this case but I don't understand how to compile a test for this. The implementation should be very similar to https://github.com/dotnet/corert/blob/635cf21aca11265ded9d78d216424bd609c052f5/src/Common/src/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs#L239

3) I cannot compile a test using ILAsm with method with same name/sign...case tested by PE Verify

Reply:
/cc @jkotas @MichalStrehovsky @jcouv

Reply:
> cannot compile a test using ILAsm with method with same name/sign...case tested by PEVerify 

Have you tried `/ERROR` command line switch for ilasm? This switch was specifically invented to allow creating invalid binaries. If there is error checking in ilasm that is getting into way, it should be ok to suppress it under the `/ERROR` switch.

Reply:
>  The implementation should be very similar to

It would be nice to figure out how to call that algorithm from here instead of duplicating it.

Reply:
> It would be nice to figure out how to call that algorithm from here (or at least part of it) instead of duplicating it.

We could use extension https://github.com/dotnet/corert/blob/master/src/Common/src/TypeSystem/Common/TypeSystemHelpers.cs#L269 that call https://github.com/dotnet/corert/blob/master/src/Common/src/TypeSystem/Common/TypeSystemContext.cs#L736 that today throw NIE. I was wondering if we can fix throw with `new MetadataVirtualMethodAlgorithm()` https://github.com/dotnet/corert/blob/master/src/Common/src/TypeSystem/Common/MetadataVirtualMethodAlgorithm.cs#L11 that's implements the logic I "copied", but It's more complete. Tests works well with it. Is there a reason for throw instead of use `MetadataVirtualMethodAlgorithm` class?

Reply:
> Is there a reason for throw

We throw from the base type system context implementation because resolving virtual methods can get weird for e.g. array types and it's better if the type system consumer (e.g. the ILVerify project) provides their own implementation.

I would recommend overriding the method in `ILVerifyTypeSystemContext` with what you're suggesting.